### ISSUE TITLE
Rule Change: Allow complex, non-name-specific RexExp strings for no-unused-vars varsIgnorePattern

### ISSUE THREAD

**Issue by D-Pow** [eslint/eslint#15078]
### What rule do you want to change?

no-unused-vars

### What change to do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

Other

### Example code

```js
function func() {}
const inlineFunc = (a, { b = 'hi', c = 'bye' }) => {};
class MyClass {
    static myStaticFunc() {
        //
    }

    static myStaticArrowFunc = () => {
        //
    };

    myFunc() {
        //
    }

    myArrowFunc = () => {
        //
    };
}
const MyJsxComponent = () => (<div/>);
```

### What does the rule currently do for this code?

Error: `func`, `inlineFunc`, `myStaticArrowFunc`, and `MyJsxComponent` are defined but never used.

### What will the rule do after it's changed?

Allow complex regex for checking the context _around_ the name to be taken into consideration rather than _only_ the name before determining if a warning/error should be thrown.

For example, I would like to allow unused functions but not unused variables. Adding the configuration below to `no-unused-vars` doesn't disable the check for functions even though it matches every single function in the above code example.

```
"varsIgnorePattern": "(?<= )(\\w+)(?=( = )?\\([^\\)]*\\)( =>)? [\\(\\{])"
```

Rather than request a new option for `ignoreFunctions`, I think a better request that would be all-encompassing of future requests would be to allow the RegExp to check the context around the variable itself so that users can implement their own RegExp to disable no-unused-vars for anything they want.

For reference, the offending line is [here](https://github.com/eslint/eslint/blob/305e14af8bd12afc01487abee5c9b0f3eaca989e/lib/rules/no-unused-vars.js#L623).

### Participation

- [X] I am willing to submit a pull request to implement this change.

### Additional comments

FYI I think implementing this change would make implementing e.g. `args.after-used`, `ignoreRestSiblings`, and `caughtErrors` easier to implement because you could just feed all the data from them into the new, e.g. `checkRegexMatch()`, utility function that would be created by course of implementing this change.

**Comment by nzakas** [comment]
> Allow complex regex for checking the context around the name to be taken into consideration rather than only the name before determining if a warning/error should be thrown.

I don’t like this approach. Regexes aren’t a good way to determine context in a language as complex as JavaScript. 

> I would like to allow unused functions but not unused variables.

I think an ignoreFunctions option is much more appropriate for your use case.

**Comment by D-Pow** [comment]
> I don’t like this approach. Regexes aren’t a good way to determine context in a language as complex as JavaScript.

Yes and no. There exist cases for both and I don't think simply dismissing one method of operation just because it isn't "a good way" is somewhat subjective and short-sighted. If someone has the knowledge to use regex, then shouldn't they be empowered to use it to the utmost?

> I think an `ignoreFunctions` option is much more appropriate for your use case.

I understand that this might possibly be more direct of an option for users, but at the end of the day, even `ignoreFunctions` requires a complicated regex/AST parsing system, right? You might as well implement it via regex (which I already provided above) and then expand upon that via `ignoreAnything: RegExp`.

**Comment by nzakas** [comment]
> I understand that this might possibly be more direct of an option for users, but at the end of the day, even ignoreFunctions requires a complicated regex/AST parsing system, right?

Not really. Any node representing a function has an type containing the word “Function”. It is fairly straightforward to look at neighboring nodes to get the needed context.

**Comment by mdjermanovic** [comment]
> `"varsIgnorePattern": "(?<= )(\\w+)(?=( = )?\\([^\\)]*\\)( =>)? [\\(\\{])"`

This pattern wouldn't find any of the following functions:

```js
const inlineFunc=() => {};

const inlineFunc = ()=>{};

const inlineFunc = a => {};

const inlineFunc = () => null;

const inlineFunc = async () => {};

const inlineFunc = (a = f()) => {};
```

These examples were not intended to show that this pattern isn't good, but to imply that regular expressions are not good for this purpose.

> I understand that this might possibly be more direct of an option for users, but at the end of the day, even `ignoreFunctions` requires a complicated regex/AST parsing system, right? You might as well implement it via regex (which I already provided above) and then expand upon that via `ignoreAnything: RegExp`.

We would never use a regex to search for functions in the source text, because it isn't a reliable way to find functions in JS code. It's parser's job to find functions in the source text and represent them in the AST, and then we can easily find them in the AST.

> If someone has the knowledge to use regex, then shouldn't they be empowered to use it to the utmost?

No, because it isn't a reliable way to analyze JS code, while ESLint provides much better ways. If anything, we could consider adding an option to specify [AST selectors](https://eslint.org/docs/developer-guide/selectors).

**Comment by D-Pow** [comment]
Apologies if I worded that in a blunt manner, but I understand what you two are saying.

> This pattern wouldn't find any of the following functions: ...

True, even I had noticed before creating the issue that stuff like `const myFunc = x => ...` weren't matched. It's not a foolproof regex, just one I whipped up quickly to use as an example.

That being said, I agree with the "ESLint provides much better ways" sentiment because even I would find that easier to use than regex. If possible, I'd love to see this configuration option added to all rules. There are others besides `no-unused-vars` that I'd like to disable in very specific circumstances without having to use comments, and the AST selectors would do just that. Granted, `no-unused-vars` is the one needing it the most.

Is this something the team would be willing to do (for `no-unused-vars` and/or all rules)?

**Comment by nzakas** [comment]
@D-Pow I’m sorry, core rules aren’t designed to catch every person’s unique preferences. The rules are there to get about 90% of the cases. If you need anything more than that, you’ll need to create your own rules. That’s exactly why plug-ins exist.

**Comment by btmills** [comment]
I agree this sounds like the perfect use case for a customized version of `no-unused-vars` in a plugin. An `ignoreFunctions` option might be reasonable in core if there were a lot of demand for it, but I don't see us adding AST selector-based ignores to rule options.

**Comment by nzakas** [comment]
We couldn’t reach consensus on this, so closing.
